contracts/src/Gateway.sol

```diff
@@ -493,199 +391,182 @@ contract Gateway is IGateway, IInitializable, IUpgradable {
+    *    /  _  \  \______   \|   | ___  __\_____  \
+    *   /  /_\  \  |     ___/|   | \  \/ / /  ____/§
+    *  /    |    \ |    |    |   |  \   / /       \
+    *  \____|__  / |____|    |___|   \_/  \_______ \
+    *          \/                                 \/
+    */
 
-    // Submit an outbound message to Polkadot, after taking fees
-    function _submitOutbound(Ticket memory ticket) internal {
-        ChannelID channelID = ticket.dest.into();
-        Channel storage channel = _ensureChannel(channelID);
+    uint256 public constant DISPATCH_OVERHEAD_GAS_V2 = 32_000;
 
-        // Ensure outbound messaging is allowed
-        _ensureOutboundMessagingEnabled(channel);
+    /// @dev Submit a message from Polkadot for verification and dispatch
+    /// @param message A message produced by the OutboundQueue pallet on BridgeHub
+    /// @param leafProof A message proof used to verify that the message is in the merkle tree committed by the OutboundQueue pallet
+    /// @param headerProof A proof that the commitment is included in parachain header that was finalized by BEEFY.
+    /// @param rewardAddress Account on BH to credit delivery rewards
+    function v2_submit(
+        InboundMessageV2 calldata message,
+        bytes32[] calldata leafProof,
+        Verification.Proof calldata headerProof,
+        bytes32 rewardAddress
+    ) external {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
 
-        // Destination fee always in DOT
-        uint256 fee = _calculateFee(ticket.costs);
+        bytes32 leafHash = keccak256(abi.encode(message));
 
-        // Ensure the user has enough funds for this message to be accepted
-        if (msg.value < fee) {
-            revert FeePaymentToLow();
+        if ($.inboundNonce.get(message.nonce)) {
+            revert IGateway.InvalidNonce();
         }
 
-        channel.outboundNonce = channel.outboundNonce + 1;
+        $.inboundNonce.set(message.nonce);
 
-        // Deposit total fee into agent's contract
-        payable(channel.agent).safeNativeTransfer(fee);
+        // Produce the commitment (message root) by applying the leaf proof to the message leaf
+        bytes32 commitment = MerkleProof.processProof(leafProof, leafHash);
 
-        // Reimburse excess fee payment
-        if (msg.value > fee) {
-            payable(msg.sender).safeNativeTransfer(msg.value - fee);
+        // Verify that the commitment is included in a parachain header finalized by BEEFY.
+        if (!_verifyCommitment(commitment, headerProof)) {
+            revert IGateway.InvalidProof();
         }
 
-        // Generate a unique ID for this message
-        bytes32 messageID = keccak256(abi.encodePacked(channelID, channel.outboundNonce));
+        bool success = v2_dispatch(message);
 
-        emit IGateway.OutboundMessageAccepted(channelID, channel.outboundNonce, messageID, ticket.payload);
+        emit IGateway.InboundMessageDispatched(message.nonce, success, rewardAddress);
```

Comment:
Would it be better to store the delivery status on chain? e.g:

```
mapping(uint64 nonce => DeliveryStatus) public nonces;
struct DeliveryStatus {
    uint64 blockNumber;
    bool result;
}
```

With the `blockNumber` it's easy for relayer to filter this event efficiently when submit the delivery proof.

Reply:
That's a good point Ron. Though there are some simplifying assumptions which I think allow us to save 20000 gas by not storing that status on-chain.

The plan I had in mind was that the relayer who calls `v2_submit` is also responsible for sending delivery proof back to BH. Since this relayer witnessed the result for `v2_submit` it can generate a proof for `InboundMessageDispatched`, and there isn't a need for additional metadata on-chain.

For other cases when the delivery proof could not be sent back immediately, we could create a command in `snowbridge-relay` that queries for `InboundMessageDispatched(rewardAddress=X)` against some public ethereum node or event indexer, and sends the delivery proofs for the matching events. 

I will make a `rewardAddress` an indexed event parameter, which allows this query to be much faster to execute.

Reply:
https://github.com/Snowfork/snowbridge/pull/1321/commits/7187bf3ef7eaffa793b269b88736ecae8b8ac7dc